spec/AuthenticationAdapters.spec.js

```diff
@@ -2160,3 +2172,1033 @@ describe('facebook limited auth adapter', () => {
+    validateAuthData: () => Promise.resolve({ doNotSave: true }),
+  };
+
+  const additionalAdapter = {
+    validateAppId: () => Promise.resolve(),
+    validateAuthData: () => Promise.resolve(),
+    policy: 'additional',
+  };
+
+  const soloAdapter = {
+    validateAppId: () => Promise.resolve(),
+    validateAuthData: () => Promise.resolve(),
+    policy: 'solo',
+  };
+
+  const challengeAdapter = {
+    validateAppId: () => Promise.resolve(),
+    validateAuthData: () => Promise.resolve(),
+    challenge: () => Promise.resolve({ token: 'test' }),
+    options: {
+      anOption: true,
+    },
+  };
+
+  const modernAdapter = {
+    validateAppId: () => Promise.resolve(),
+    validateSetUp: () => Promise.resolve(),
+    validateUpdate: () => Promise.resolve(),
+    validateLogin: () => Promise.resolve(),
+  };
+
+  const modernAdapter2 = {
+    validateAppId: () => Promise.resolve(),
+    validateSetUp: () => Promise.resolve(),
+    validateUpdate: () => Promise.resolve(),
+    validateLogin: () => Promise.resolve(),
+  };
+
+  const wrongAdapter = {
+    validateAppId: () => Promise.resolve(),
+  };
+
+  const headers = {
+    'Content-Type': 'application/json',
+    'X-Parse-Application-Id': 'test',
+    'X-Parse-REST-API-Key': 'rest',
+  };
+
+  it('should ensure no duplicate auth data id after before save', async () => {
+    await reconfigureServer({
+      auth: { baseAdapter },
+      cloud: () => {
+        Parse.Cloud.beforeSave('_User', async request => {
+          request.object.set('authData', { baseAdapter: { id: 'test' } });
+        });
+      },
+    });
+
+    const user = new Parse.User();
+    await user.save({ authData: { baseAdapter: { id: 'another' } } });
+    await user.fetch({ useMasterKey: true });
+    expect(user.get('authData')).toEqual({ baseAdapter: { id: 'test' } });
+
+    const user2 = new Parse.User();
+    await expectAsync(
+      user2.save({ authData: { baseAdapter: { id: 'another' } } })
+    ).toBeRejectedWithError('this auth is already used');
+  });
+
+  it('should ensure no duplicate auth data id after before save in case of more than one result', async () => {
+    await reconfigureServer({
+      auth: { baseAdapter },
+      cloud: () => {
+        Parse.Cloud.beforeSave('_User', async request => {
+          request.object.set('authData', { baseAdapter: { id: 'test' } });
+        });
+      },
+    });
+
+    const user = new Parse.User();
+    await user.save({ authData: { baseAdapter: { id: 'another' } } });
+    await user.fetch({ useMasterKey: true });
+    expect(user.get('authData')).toEqual({ baseAdapter: { id: 'test' } });
+
+    let i = 0;
+    const originalFn = Auth.findUsersWithAuthData;
+    spyOn(Auth, 'findUsersWithAuthData').and.callFake((...params) => {
+      // First call is triggered during authData validation
+      if (i === 0) {
+        i++;
+        return originalFn(...params);
+      }
+      // Second call is triggered after beforeSave.
```

Comment:
can you use equal line length in this comment block to make it easier to read, also rephrase, it's hard to understand without punctuation.